onmt/decoders/decoder.py

```diff
@@ -168,12 +169,28 @@ def _fix_enc_hidden(hidden):
             return hidden
 
         if isinstance(encoder_final, tuple):  # LSTM
-            return RNNDecoderState(self.hidden_size,
-                                   tuple([_fix_enc_hidden(enc_hid)
-                                          for enc_hid in encoder_final]))
+            self.state["hidden"] = tuple([_fix_enc_hidden(enc_hid)
+                                          for enc_hid in encoder_final])
         else:  # GRU
-            return RNNDecoderState(self.hidden_size,
-                                   _fix_enc_hidden(encoder_final))
+            self.state["hidden"] = (_fix_enc_hidden(encoder_final), )
+
+        # Init the input feed.
+        batch_size = self.state["hidden"][0].size(1)
+        h_size = (batch_size, self.hidden_size)
+        self.state["input_feed"] = \
```

Comment:
The input feeding logic should only be part of `InputFeedRNNDecoder`. You can add overrides in this class, call the base methods and add the input feed state.

Reply:
Let's make another PR after this one that removes `update_state`. It does not appear useful over directly updating the state fields.

Reply:
yes, correct.